### ISSUE TITLE
Sub-object values appear to not be read back from CKRecord

### ISSUE THREAD

**Issue by tbaggett** [evermeer/EVCloudKitDao#52]
I am taking advantage of your recently-added support for subobjects to be written and read from iCloud. I have local caching disabled to avoid the current failure to restore all properties via NSCoder.

I have the following class/property structure:

    class PerformanceContent: 
        var options: PerformanceOptions:
            var playback: PlaybackOptions:
                var endlessPlayback: CloudBool
                    var state: CloudBoolState (an integer-based enumeration)

This nested class structure results in this field in the PerformanceContent CKRecord:

    options__playback__endlessPlayback__state

When I assign a different enumeration value to the state property, it is written as expected. I see the correct value when I inspect the record in the CloudKit dashboard. However, it is always set to its initially-assigned enumeration value (Unassigned, -1) when I shut down and restart the app. 

Note that the top-level PerformanceContent class (which the CKRecord is based on) doesn't contain any immediate properties, meaning strings, ints, etc. Every field in the record is a sub-object. I mention this in case it gives a clue as to what is causing this issue.

Here is the source for my CloudBool class. I expected the setValue:forUndefinedKey method to be called to restore the state property's value since it is an enumeration, but a breakpoint on the method never gets hit. 

	public class CloudBool: TBCloudKitDataObject {

		public enum CloudBoolState: Int, EVRawInt {
			case Unassigned = -1, False = 0, True = 1
		
			public init() { self = Unassigned }
		
			public init(value: Bool?) {
				self.init()
			
				assignOptionalBool(value)
			}
		
			public mutating func assignOptionalBool(value: Bool?) {
				if let val = value {
					if val {
						self = .True
					} else {
						self = .False
					}
				} else {
					self = .Unassigned
				}
			}
		
			public mutating func assignRawValue(rawValue: Int) {
				if let val = CloudBoolState(rawValue: rawValue) {
					self = val
				}
			}
		
			public func asBool() -> Bool {
				switch self {
				case .True:
					return true
				default:
					return false
				}
			}
		
			public func asOptionalBool() -> Bool? {
				switch self {
				case .Unassigned:
					return nil
				case .False:
					return false
				case .True:
					return true
				}
			}
		}
	
		public var state: CloudBoolState = .Unassigned
	
		public override func setValue(value: AnyObject!, forUndefinedKey key: String) {
			switch key {
			case "state":
				state.assignRawValue(value as! Int)
			default:
				super.setValue(value, forUndefinedKey: key)
			}
		}
	}

**Comment by evermeer** [comment]
I have looked through your code and I can't see anything wrong with it. You are using the EVRawInt and setValue forUndefinedKey as you should for an enum.

Could you debug this method:
https://github.com/evermeer/EVCloudKitDao/blob/14c9da0b7ff7d93ce0cc6376062c453b0439c2b3/AppMessage/AppMessage/CloudKit/EVCloudKitDao.swift#L1074-1074

especially the else case after the split for __ 
The list repeat in the loop should set the state in the dictionary.

**Comment by tbaggett** [comment]
Sure, I will step through that and see what is going on. I will also test the sorting stuff that you added today. I was tied up with this yesterday and didn't make it to that.

**Comment by tbaggett** [comment]
So I stepped through the CKRecordToDictionary method as you asked. The dictionary that is created by that method is correct. It creates all the sub-object dictionaries and correctly assigns their values. I see that method was called by fromCKRecord, which uses NSKeyedArchiver to encode fields, then returns the resulting object to queryRecords, this version specifically:

    internal func queryRecords<T:EVCloudKitDataObject>(type:T, query: CKQuery, completionHandler: (results: [T], isFinished: Bool) -> Bool, errorHandler:((error: NSError) -> Void)? = nil)

In queryRecords, this code casts the EVCloudKitDataObject to my class and appends it to the result set using this code:

        operation.recordFetchedBlock = { record in
            if let parsed = self.fromCKRecord(record) as? T  {
                results.append(parsed)
            }

Where should the setValue:forUndefinedKey method call be happening in this sequence?

**Comment by tbaggett** [comment]
Okay I did a bit of reading and realized I misunderstood the source of setValue:forUndefinedKey. I didn't realize it is a part of the NSCoding and NSKeyedArchiver and as such should be called by them. I thought it was part of EVReflection. Apologies for that.

I also found through further testing that enums don't behave the same as classes, despite Apple documentation vaguely saying they do. For example, what I considered to be the enum's default initializer, init(), isn't called if you declare an enum property and assign a value to it, i.e, myEnum = .myState. It is only called if you explicitly call it, i.e., myEnum = MyEnum(). You also can't specify that an initializer is required in an enum. Doing so results in a compile error.

Long story short, I've moved the functionality I had in the CloudBoolState enum out and placed it in the CloudBool class. It really makes more sense when I considered how it is used anyway. I am also going to refactor all of my data classes to have simple DTO classes that I will use with EVCloudKitDao instead of trying to have one consolidated class for each CKRecord type that also contains supporting logic. Hopefully doing so will make everything work as intended.
